fem/gslib.cpp

```diff
@@ -184,35 +181,273 @@ void FindPointsGSLIB::Interpolate(Array<unsigned int> &codes,
+      const IntegrationRule &ir     = fes->GetFE(0)->GetNodes();
+      const int dof_cnt = ir.GetNPoints();
+
+      node_vals.SetSize(NE * dof_cnt);
+
+      const TensorBasisElement *tbe =
+         dynamic_cast<const TensorBasisElement *>(fes->GetFE(0));
+      MFEM_VERIFY(tbe != NULL, "TensorBasis FiniteElement expected.");
+      const Array<int> &dof_map = tbe->GetDofMap();
+
+      int pt_id = 0;
+      Vector vals_el;
+      for (int i = 0; i < NE; i++)
+      {
+         gf_in.GetValues(i, ir, vals_el);
+         for (int j = 0; j < dof_cnt; j++)
+         {
+            node_vals(pt_id++) = vals_el(dof_map[j]);
+         }
+      }
+   }
+   else if (gt == Geometry::TRIANGLE || gt == Geometry::TETRAHEDRON ||
+            gt == Geometry::PRISM)
+   {
+      const int dof_cnt = ir_simplex->GetNPoints();
+      node_vals.SetSize(NE * dof_cnt);
+
+      int pt_id = 0;
+      Vector vals_el;
+      for (int j = 0; j < NE; j++)
+      {
+         gf_in.GetValues(j, *ir_simplex, vals_el);
+         for (int i = 0; i < dof_cnt; i++)
+         {
+            node_vals(pt_id++) = vals_el(i);
+         }
+      }
+   }
+   else
+   {
+      MFEM_ABORT("Element type not currently supported.");
+   }
+}
+
+void FindPointsGSLIB::GetQuadHexNodalCoordinates()
+{
+   const GridFunction *nodes     = mesh->GetNodes();
    const FiniteElementSpace *fes = nodes->FESpace();
-   const IntegrationRule &ir = fes->GetFE(0)->GetNodes();
 
-   const int NE = mesh->GetNE(), dof_cnt = ir.GetNPoints();
-   node_vals.SetSize(NE * dof_cnt);
+   const int NE      = mesh->GetNE(),
+             dof_cnt = fes->GetFE(0)->GetDof(),
+             pts_cnt = NE * dof_cnt;
+   gsl_mesh.SetSize(dim * pts_cnt);
 
    const TensorBasisElement *tbe =
       dynamic_cast<const TensorBasisElement *>(fes->GetFE(0));
+   MFEM_VERIFY(tbe != NULL, "TensorBasis FiniteElement expected.");
    const Array<int> &dof_map = tbe->GetDofMap();
 
+   DenseMatrix pos(dof_cnt, dim);
+   Vector posV(pos.Data(), dof_cnt * dim);
+   Array<int> xdofs(dof_cnt * dim);
+
    int pt_id = 0;
-   Vector vals_el;
    for (int i = 0; i < NE; i++)
    {
-      gf_in.GetValues(i, ir, vals_el);
+      fes->GetElementVDofs(i, xdofs);
+      nodes->GetSubVector(xdofs, posV);
       for (int j = 0; j < dof_cnt; j++)
       {
-         node_vals(pt_id++) = vals_el(dof_map[j]);
+         for (int d = 0; d < dim; d++)
+         {
+            gsl_mesh(pts_cnt * d + pt_id) = pos(dof_map[j], d);
+         }
+         pt_id++;
+      }
+   }
+}
+
+void FindPointsGSLIB::GetSimplexNodalCoordinates()
+{
+   const FiniteElement *fe   = mesh->GetNodalFESpace()->GetFE(0);
+   const Geometry::Type gt   = fe->GetGeomType();
+   const GridFunction *nodes = mesh->GetNodes();
+   Mesh *meshsplit           = NULL;
+   const int NE              = mesh->GetNE();
+   int NEsplit;
+
+   // Split the reference element into a reference submesh of quads or hexes.
+   if (gt == Geometry::TRIANGLE)
```

Comment:
I wonder if this could be of use in other parts of the code? We don't have to change it now, but we may consider moving it into an auxiliary Mesh function if there are other uses...